### ISSUE TITLE
Upper limit of the number of fractional bits `f`

### ISSUE THREAD

**Issue by kimikage** [JuliaMath/FixedPointNumbers.jl#155]
In theory, the number of fractional bits `f` can be countlessly large, but what is the actual upper limit?
I assumed that it is `8sizeof(T)-1` for `Fixed{T}`, and `8sizeof(T)` for `Normed{T}`.

However, the larger numbers are also used actually. 
https://github.com/JuliaMath/FixedPointNumbers.jl/blob/7ad0f0c460961024191f672332fcf3d7cef655d1/test/fixed.jl#L63

https://github.com/JuliaMath/FixedPointNumbers.jl/blob/7ad0f0c460961024191f672332fcf3d7cef655d1/test/fixed.jl#L96

https://github.com/JuliaMath/FixedPointNumbers.jl/blob/7ad0f0c460961024191f672332fcf3d7cef655d1/test/fixed.jl#L110

https://github.com/JuliaMath/FixedPointNumbers.jl/blob/7ad0f0c460961024191f672332fcf3d7cef655d1/test/fixed.jl#L138

Perhaps some functions do not work properly with large `f`, due to implicit upper limits.

**Edit:**
I intend to support `Fixed{T,f}` where `f == 8sizeof(T)`, i.e. throw no exceptions. However, they have no typealias and will not be tested enough.

**Comment by kimikage** [comment]
I also have some concerns about the type of `f`.

> BTW, I think there is a problem in interoperability with specialization using integer literal parameters.
> ```julia
> julia> Normed{UInt32,Int64(1)} === Normed{UInt32,Int32(1)}
> false
> 
> julia> Normed{UInt32,Int64(1)} == Normed{UInt32,Int32(1)}
> false
> ```
> 
> I don't know how serializers (e.g. JLD2) do (de-)serialization.

_Originally posted by @kimikage in https://github.com/JuliaMath/FixedPointNumbers.jl/issues/129#issuecomment-554701260_

I think `f <: Integer` is fine except for the ambiguity associated with omitting constraints, but I think `isa(f, Int)` is problematic as a constraint.
https://github.com/JuliaMath/FixedPointNumbers.jl/blob/08c7452754b15b2a214bdb844c8eb5534cb4d09d/src/normed.jl#L8

**Comment by timholy** [comment]
I think those test cases are broken, we should probably change them.

> isa(f, Int) is problematic as a constraint.

Why?

**Comment by kimikage** [comment]
> I think those test cases are broken, we should probably change them.

It is reasonable to consider them broken, but we need to be careful if they have been actually used. Should we add a depwarn as usual?
This decision affects the implementation of `floor(::Fixed)`. I have implemented the code which may throws the exception.

> Why?

You will not be able to construct `Normed{::T, ::Int32}` on the 64-bit version, nor  `Normed{::T, ::Int64}` on the 32-bit version.
If we can construct them, we can provide the methods for conversions, but if we can't, we can't even convert them.

**Comment by timholy** [comment]
> You will not be able to construct Normed{::T, ::Int32} on the 64-bit version, nor Normed{::T, ::Int64} on the 32-bit version.

I'm not sure that's a problem. In the end types get identified as a pointer (see `jltypes.c` in Julia's `src/` directory) and of course the compiler drops the types wherever possible, so a type-parameter is only needed essentially as a UUID. Is there a genuine use-case for non-`Int` `f`?

**Comment by kimikage** [comment]
I've been discussing the interoperability between the 32-bit and 64-bit systems since https://github.com/JuliaMath/FixedPointNumbers.jl/issues/129#issuecomment-554701260. If it doesn't matter, it's all fine.

**Comment by timholy** [comment]
Now I see what you mean. This really seems like something JLD2 should handle, and not be a problem that individual packages have to worry about. IMO any time there's an integer type parameter it should just be `::Int`.

**Comment by kimikage** [comment]
It is definitely a good thing that JLD2 can handle it well. However, JLD2 is not the only serializer, and FixedPointNumbers is a somewhat famous but ordinary package, and not included in the standard library.

I agree with the principle. The dimension of `Array` also follows the principle. However, when `f` is not `Int`, there should be some reasons. I don't know whether the reasons are worth breaking the principle.

In any case, even if we restrict `f` into `Int`, we need to implement functions to work whether `f` is an `Int32` or `Int64`.

**Comment by kimikage** [comment]
Is the following customization really useful? (I skeptically wrote this. )
```julia
function deserialize(s::AbstractSerializer, ::Type{Normed{T,f}}) where {T, f}
    tag = read(s.io, UInt8)
    reinterpret(Normed{T, Int(f)}, read(s.io, T))
end
```
One of the troubles of customizing `Serialization` is that container types (e.g. arrays) are specialized.

From an implementation point of view, I think it is simpler to use the promotion mechanism.

**Comment by timholy** [comment]
Hard to say in the abstract. You've probably seen what happens right now, but I haven't poked at this issue. If you need input, a set of failing tests would be a good starting point. (I don't have access to a 32-bit machine, however.)

**Comment by kimikage** [comment]
> I don't have access to a 32-bit machine, however.

I don't want to add `isa(f, Int)` to the constructors "for now", so we can reproduce the types for 32-bit systems on 64-bit systems with explicit notations like `Normed{UInt8, Int32(8)}`.

BTW, I also think AppVeyor supports the 32-bit Windows build. Travis-ci also supports the 32-bit Linux build, but I have never tried it.
cf. https://discourse.julialang.org/t/32-bit-linux-julia-builds-are-now-available-on-travis-ci/27786

I don't think `isa(f, Int) == false` is a fatal problem, so I'd like to tackle it after the upcoming set of fixes has been completed.

**Comment by timholy** [comment]
I'm in no hurry, but I do think forcing `Int` is a useful normalization:
```julia
julia> using FixedPointNumbers

julia> T1 = Normed{UInt8,8}
Normed{UInt8,8}

julia> T2 = Normed{UInt8,Int32(8)}
Normed{UInt8,8}

julia> T1 == T2
false

julia> x = T2(0.4)
0.4N0f8

julia> isa(x, N0f8)
false
```
That's *really* confusing.

**Comment by kimikage** [comment]
> That's *really* confusing.

I completely agree. I also think "forcing" `Int` is useful. The problem is the way to force. As you know, `isa(f, Int) || error()` explains what happened, but does not help further to handle data. The matter is that the targets are constructors, not factory methods.

**Comment by kimikage** [comment]
PR #159 was merged, and I submitted a new issue (#162). So, let's continue the discussion there.
